poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+ */
+func (p *Promise) sign(i int, gKeyPair *config.KeyPair, msg []byte) *PromiseSignature {
+	set := anon.Set{gKeyPair.Public}
+	sig := anon.Sign(gKeyPair.Suite, random.Stream, msg, set, nil, 0,
+		gKeyPair.Secret)
+	return new(PromiseSignature).init(gKeyPair.Suite, sig)
+}
+
+/* An internal helper function, verifies a signature is from a given insurer.
+ *
+ * Arguments
+ *    i   = the index of the insurer in the insurers list
+ *    sig = the PromiseSignature object containing the signature
+ *    msg = the message that was signed
+ *
+ * Return
+ *   an error if the signature is malformed, nil otherwise.
+ */
+func (p *Promise) verifySignature(i int, sig *PromiseSignature, msg []byte) error {
+	if i < 0 || i >= p.n {
+		return errors.New("Invalid index. Expected 0 <= i < n")
+	}
+	if sig.signature == nil {
+		return errors.New("Nil PromiseSignature")
+	}
+	set := anon.Set{p.insurers[i]}
+	_, err := anon.Verify(sig.suite, msg, set, nil, sig.signature)
+	return err
+}
+
+/* A public wrapper function for sign, produces a signature for a given insurer.
+ * Insurers should use this function to produce a PromiseSignature to express
+ * approval of a Promise.
+ *
+ * Arguments
+ *    i         = the index of the insurer's share
+ *    gKeyPair  = the long term public/private keypair of the insurer.
+ *
+ * Return
+ *   A PromiseSignature object with the signature.
+ *
+ * Note
+ *   It is wise to first call VerifyShare to verify the insurer's share before
+ *   signing on the promise.
+ */
+func (p *Promise) Sign(i int, gKeyPair *config.KeyPair) *PromiseSignature {
+	return p.sign(i, gKeyPair, sigMsg)
+}
+
+/* Verifies a signature from a given insurer
+ *
+ * Arguments
+ *    i   = the index of the insurer in the insurers list
+ *    sig = the PromiseSignature object containing the signature
+ *
+ * Return
+ *   an error if the promise is malformed, nil otherwise.
+ */
+func (p *Promise) VerifySignature(i int, sig *PromiseSignature) error {
+	return p.verifySignature(i, sig, sigMsg)
+}
+
+/* Reveals the secret share that the insurer has been protecting. The insurer
+ * should call this function on behalf of a client after verifying that the
+ * Promiser is non-responsive.
+ *
+ * Arguments
+ *    i        = the index of the insurer
+ *    gkeyPair = the long-term keypair of the insurer
+ *
+ * Return
+ *   the revealed private share
+ */
+func (p *Promise) RevealShare(i int, gKeyPair *config.KeyPair) abstract.Secret {
+	diffieBase := p.keySuite.Point().Mul(p.pubKey, gKeyPair.Secret)
+	share := p.diffieHellmanDecrypt(p.secrets[i], diffieBase)
+	return share
+}
+
+/* Verify that a revealed share is properly formed. This should be called by
+ * clients or others who request an insurer to reveal its shared secret.
+ *
+ * Arguments
+ *    i     = the index of the share
+ *    share = the share to validate.
+ *
+ * Return
+ *   Whether the secret is valid
+ */
+func (p *Promise) VerifyRevealedShare(i int, share abstract.Secret) error {
```

Comment:
VerifyRevealedShare is used to protect against malicious insurers. Although the Promise could be valid, a malicious insurer could give a client a bad share. Hence, it would not be able to recover the secret.

The function provides a way for clients to ensure that an insurer has given it the proper share.